protocol/app/upgrades/v5.2.0/upgrade_container_test.go

```diff
@@ -0,0 +1,165 @@
+//go:build all || container_test
+
+package v_5_2_0_test
+
+import (
+	"testing"
+
+	"github.com/cosmos/gogoproto/proto"
+	v_5_2_0 "github.com/dydxprotocol/v4-chain/protocol/app/upgrades/v5.2.0"
+	"github.com/dydxprotocol/v4-chain/protocol/testing/containertest"
+	"github.com/dydxprotocol/v4-chain/protocol/testutil/constants"
+	feetierstypes "github.com/dydxprotocol/v4-chain/protocol/x/feetiers/types"
+	vaulttypes "github.com/dydxprotocol/v4-chain/protocol/x/vault/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestStateUpgrade(t *testing.T) {
+	testnet, err := containertest.NewTestnetWithPreupgradeGenesis()
+	require.NoError(t, err, "failed to create testnet - is docker daemon running?")
+	err = testnet.Start()
+	require.NoError(t, err)
+	defer testnet.MustCleanUp()
+	node := testnet.Nodes["alice"]
+	nodeAddress := constants.AliceAccAddress.String()
+
+	preUpgradeSetups(node, t)
+	preUpgradeChecks(node, t)
+
+	err = containertest.UpgradeTestnet(nodeAddress, t, node, v_5_2_0.UpgradeName)
+	require.NoError(t, err)
+
+	postUpgradeChecks(node, t)
+}
+
+/*** Preupgrade Setup ***/
+func preUpgradeSetups(node *containertest.Node, t *testing.T) {
+	preUpgradeSetupVaults(node, t)
+}
+
+func preUpgradeSetupVaults(node *containertest.Node, t *testing.T) {
+	// Query vault activation threshold quote quantums.
+	params := &vaulttypes.QueryParamsResponse{}
+	resp, err := containertest.Query(
+		node,
+		vaulttypes.NewQueryClient,
+		vaulttypes.QueryClient.Params,
+		&vaulttypes.QueryParamsRequest{},
+	)
+	require.NoError(t, err)
+	err = proto.UnmarshalText(resp.String(), params)
+	require.NoError(t, err)
+
+	// Wait until subaccount transfers are enabled.
+	err = node.WaitUntilBlockHeight(51)
+
+	// Deposit to vaults.
+	require.NoError(t, containertest.BroadcastTx(
+		node,
+		&vaulttypes.MsgDepositToVault{
+			VaultId:       &constants.Vault_Clob_0,
+			SubaccountId:  &constants.Alice_Num0,
+			QuoteQuantums: params.Params.ActivationThresholdQuoteQuantums,
+		},
+		constants.AliceAccAddress.String(),
+	))
+	require.NoError(t, containertest.BroadcastTx(
+		node,
+		&vaulttypes.MsgDepositToVault{
+			VaultId:       &constants.Vault_Clob_1,
+			SubaccountId:  &constants.Bob_Num0,
+			QuoteQuantums: params.Params.ActivationThresholdQuoteQuantums,
+		},
+		constants.BobAccAddress.String(),
+	))
+	require.NoError(t, containertest.BroadcastTx(
+		node,
+		&vaulttypes.MsgDepositToVault{
+			VaultId: &vaulttypes.VaultId{
+				Type:   vaulttypes.VaultType_VAULT_TYPE_CLOB,
+				Number: 17,
+			},
+			SubaccountId:  &constants.Carl_Num0,
+			QuoteQuantums: params.Params.ActivationThresholdQuoteQuantums,
+		},
+		constants.CarlAccAddress.String(),
+	))
+
+	err = node.Wait(2)
+	require.NoError(t, err)
+}
+
+/*** Preupgrade Check ***/
+func preUpgradeChecks(node *containertest.Node, t *testing.T) {
+	// Add test for your upgrade handler logic below
+}
+
+/*** Postupgrade Check ***/
+func postUpgradeChecks(node *containertest.Node, t *testing.T) {
```

Comment:
nit: Check that vault addresses were added to the vault state holding addresses.

Reply:
i don't know if that's accessible through `node` / container tests. Instead, I opted to check whether every vault belongs to the best fee tier.

Reply:
non-blocking nit: Check that they aren't part of the best fee tier pre-upgrade?